File: corporate-hotel-booking/src/booking/booking_policy_service.py
from booking.booking_policy_repository import BookingPolicyRepository
from hotel.hotel import RoomType


class BookingPolicyService:
    def __init__(self, booking_policy_repository: BookingPolicyRepository):
        self.__booking_policy_repository = booking_policy_repository

    def set_company_policy(self, company_id: str, room_types: list[RoomType]) -> None:
        raise NotImplementedError()

    def set_employee_policy(self, company_id: str, room_types: list[RoomType]) -> None:
        raise NotImplementedError()

    def is_booking_allowed(self, employee_id: str, room_type: RoomType) -> bool:
        raise NotImplementedError()

File: corporate-hotel-booking-java/src/main/java/com/github/ambersariya/booking/BookingPolicyService.java
package com.github.ambersariya.booking;

import com.github.ambersariya.employee.EmployeeRepository;
import com.github.ambersariya.hotel.RoomType;

import java.util.List;

public class BookingPolicyService {
    private final BookingPolicyRepository bookingPolicyRepository;
    private final EmployeeRepository employeeRepository;

    public BookingPolicyService(BookingPolicyRepository bookingPolicyRepository, EmployeeRepository employeeRepository) {
        this.bookingPolicyRepository = bookingPolicyRepository;
        this.employeeRepository = employeeRepository;
    }

    public void setCompanyPolicy(int companyId, List<RoomType> companyRoomTypes) {
        var companyPolicy = new CompanyPolicy(companyId, companyRoomTypes);
        bookingPolicyRepository.save(companyPolicy);
    }

    public void setEmployeePolicy(int employeeId, List<RoomType> employeeRoomTypes) {
        var employeePolicy = new EmployeePolicy(employeeId, employeeRoomTypes);
        bookingPolicyRepository.save(employeePolicy);
    }

    public boolean isBookingAllowed(int employeeId, RoomType roomType) {
        var employeePolicy = bookingPolicyRepository.findEmployeePolicyBy(employeeId);
        if (employeePolicy != null) {
            return employeePolicy.roomTypes().contains(roomType);
        }

        var employee = employeeRepository.findById(employeeId);
        var companyPolicy = bookingPolicyRepository.findCompanyPolicyBy(employee.companyId());
        if (companyPolicy != null) {
            return companyPolicy.roomTypes().contains(roomType);
        }

        return true;
    }
}
